csrc/xpu_kernels.cpp

```diff
@@ -0,0 +1,314 @@
+  // power-of-2)
+  local_abs_max = absmax[(base_idx + local_idx) >>
+                         (31 - std::countl_zero<unsigned int>(blocksize))];
+
+  if (local_idx + NUM_PER_TH < valid_items_load) {
+    reinterpret_cast<sycl::vec<uint8_t, NUM_PER_TH>(&)[NUM_PER_TH]>(qvals)[0] =
+        reinterpret_cast<sycl::vec<uint8_t, NUM_PER_TH> *>(
+            A)[(base_idx + local_idx) / NUM_PER_TH];
+  } else {
+#pragma unroll NUM_PER_TH
+    for (int i = 0; i < NUM_PER_TH; i++) {
+      if (local_idx + i < valid_items_load) {
+        qvals[i] = A[base_idx + local_idx + i];
+      } else {
+        qvals[i] = (uint8_t)0;
+      }
+    }
+  }
+
+  switch (DATA_TYPE) {
+  case General8bit:
+#pragma unroll NUM_PER_TH
+    for (int j = 0; j < NUM_PER_TH; j++)
+      vals[j] = code[qvals[j]] * local_abs_max;
+    break;
+  case FP4:
+#pragma unroll NUM_PER_TH
+    for(int j = 0; j < NUM_PER_TH; j++) {
+      vals[j * 2] = dDequantizeFP4Tree(qvals[j] >> 4) * local_abs_max;
+      vals[j * 2 + 1] = dDequantizeFP4Tree(qvals[j] & 0x0F) * local_abs_max;
+    }
+    break;
+  case NF4:
+#pragma unroll NUM_PER_TH
+    for (int j = 0; j < NUM_PER_TH; j++) {
+      vals[j * 2] = dDequantizeNF4(qvals[j] >> 4) * local_abs_max;
+      vals[j * 2 + 1] = dDequantizeNF4(qvals[j] & 0x0F) * local_abs_max;
+    }
+    break;
+  }
+
+  const int local_dst_size = (DATA_TYPE > 0) ? NUM_PER_TH * 2 : NUM_PER_TH;
+  int local_dst_idx = (DATA_TYPE > 0) ? local_idx * 2 : local_idx;
+  if (local_dst_size < valid_items_store) {
+    reinterpret_cast<sycl::vec<T, local_dst_size> *>(
+        out)[(((DATA_TYPE > 0) ? base_idx * 2 : base_idx) + local_dst_idx) /
+             local_dst_size] =
+        reinterpret_cast<sycl::vec<T, local_dst_size>(&)[local_dst_size]>(
+            vals)[0];
+  } else {
+#pragma unroll NUM_PER_TH
+    for (int i = 0; i < local_dst_size; i++) {
+      if (i < valid_items_store) {
+        out[((DATA_TYPE > 0) ? base_idx * 2 : base_idx) + local_dst_idx + i] =
+            vals[i];
+      }
+    }
+  }
+}
+
+#define num_values_4bit 32
+template <typename T, int THREADS, int BITS, int SUBG_SIZE>
+SYCL_EXTERNAL void kgemv_4bit_inference_kernel(
+    int M, int N, int K, T *A, unsigned char *B, float *absmax,
+    const float *datatype, T *out, int lda, int ldb, int ldc, int blocksize,
+    sycl::local_ptr<T> quant_map, sycl::nd_item<1> &item) {
+  size_t idx = item.get_local_id();
+  const int sg_idx = idx / SUBG_SIZE;
+  const int sg_lane = idx % SUBG_SIZE;
+  const int row_B =
+      (THREADS / SUBG_SIZE) * item.get_group().get_group_id() + sg_idx;
+  const int offset_B = ldb * row_B;
+  const int num_values_8bit = num_values_4bit / 2;
+  float local_C = 0.0f;
+
+  unsigned char local_B_4bit[num_values_8bit];
+  T local_B[num_values_4bit / 4];
+  T local_A[num_values_4bit / 4];
+  T local_absmax = T(0.0f);
+
+  if (idx < 16) {
+    quant_map[idx] = T(datatype[idx]);
+  }
+
+  item.barrier(sycl::access::fence_space::local_space);
```

Comment:
Using group_barrier instead for SLM memory fence.

Reply:
No, we need subgroup barrier here for the functionality requirement, because it's reduced over subgroup instead of group.